### PR TITLE
Add support for storing the optimization targets and direction of an experiment

### PR DESCRIPTION
**Pull request by bpkroth**
This PR is useful for mlos-viz and dabl wrapper (#624) to be able to automatically graph the results for a given optimization target, for instance via something like the following:

```python
for opt_target in exp.objectives:
    dabl.plot(exp.results, opt_target)
```

Since the prior efforts on capturing this data in the Trial metadata are somewhat problematic (allow conflicting changes between runs of an experiment, don't support multi-objective), we extend them to also store values directly as a part of the Experiment, which is a somewhat more appropriate location.  Upon retrieval, an attempt is also made to merge the two data sources for backwards compatibility.

This PR does not enforce strictness on that metadata, but future versions could (e.g., disallow resuming an Experiment if it looks like the objective targets have changed.  In that case the prior Trial results can potentially still be used to prewarm a new Experiment's optimizer).

### CODE REVIEW

**Comment by motus** [comment]
wait, we already store that data in the `trial_param` table - see https://github.com/microsoft/MLOS/blob/2032d7908fd3dfd748cf30c331013cd4d0f66541/mlos_bench/mlos_bench/run.py#L120

you can add more key/value pairs to store with each trial there

**Comment by bpkroth** [comment]
> wait, we already store that data in the `trial_param` table - see
> 
> https://github.com/microsoft/MLOS/blob/2032d7908fd3dfd748cf30c331013cd4d0f66541/mlos_bench/mlos_bench/run.py#L120-L124
> 
> 
> you can add more key/value pairs to store with each trial there

We had a conversation about this offline.  I think the conclusion was that:
1. The optimization configuration is really a property of the Experiment and not the Trial, so if something about the Experiment's optimization goal's configuration changes, we should really require a new Experiment, though previously we didn't store or enforce this.
2. That doesn't preclude reusing *some* TrialData from a previous Experiment to pre-warm a new one.
3. However, storing metadata about the Optimization goals used during a given Trial, which potentially redundant, may still be useful.
    Though, with that in mind, we may also want to store other parameter data used too ...

I think with that in mind, I will still pursue this PR and a future one can enforce the strictness aspects of not changing things, perhaps after we make it easier to merge other "related" TrialData (though capturing which those are might itself be tricky).

File: mlos_bench/mlos_bench/storage/base_storage.py
Comment (bpkroth) [inline mlos_bench/mlos_bench/storage/base_storage.py:132]: Moving some of these args from the sql class to the base class since I think they make sense for any implementation.

File: mlos_bench/mlos_bench/tests/storage/trial_data_test.py
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/storage/trial_data_test.py:9]: This is a stub for now.  There were no tests for TrialData loading previously.  I didn't really change that code yet, so didn't add tests for it yet, but wanted to note that we need some while I was adding some for the ExperimentData class.

File: mlos_bench/mlos_bench/storage/sql/experiment_data.py
Comment (motus) [inline mlos_bench/mlos_bench/storage/sql/experiment_data.py:51]: I wonder if the order of optimization targets is important? Maybe we need an extra column like "order" or "priority" in the `objectives` table?

File: mlos_bench/mlos_bench/storage/base_storage.py
Comment (motus) [inline mlos_bench/mlos_bench/storage/base_storage.py:87]: I think that optimization direction can be nullable - e.g. when we run a standalone benchmark. (we should still have the optimization target, though, because even in a standalone benchmark we still measure something)

File: mlos_bench/mlos_bench/storage/sql/experiment_data.py
Comment (bpkroth) [reply mlos_bench/mlos_bench/storage/sql/experiment_data.py:51]: In true multi-objective, no, they should be equally explored by the optimizer, but it's not a bad thing to include for future support.
How about just an optional floating point "weight"?

Comment (bpkroth) [reply mlos_bench/mlos_bench/storage/sql/experiment_data.py:52]: I added the column for now, but am leaving it not fully supported in the rest of the APIs just yet as we don't fully support multi-objective yet anyways and adding weights would change a lot of these APIs more that I need for the initial dabl one sooner.

File: mlos_bench/mlos_bench/storage/base_experiment_data.py
Comment (motus) [inline mlos_bench/mlos_bench/storage/base_experiment_data.py:67]:
```suggestion
        objectives : Dict[str, objective]
```

**Review by motus — APPROVED** [review]
Looks good! I leave it up to you to keep the optimization direction nullable or roll back to the non-nullable variant

**Pull request merged**
